src/Arrays/MPIStateArrays.jl

```diff
@@ -334,149 +334,192 @@ end
     dest
 end
 
-"""
-    post_Irecvs!(Q::MPIStateArray)
-
-posts the `MPI.Irecv!` for `Q`
-"""
-function post_Irecvs!(Q::MPIStateArray)
-    nnabr = length(Q.nabrtorank)
-    transfer = get_transfer(Q.recv_buffer)
 
-    for n in 1:nnabr
-        # If this fails we haven't waited on previous recv!
-        @assert Q.recvreq[n].buffer == nothing
+# The following `__no_overlap_*` functions exist to support the old DG balance
+# law where we still use GPUifyLoops.  They should NOT be used in new code.
+function __no_overlap_begin_ghost_exchange!(Q::MPIStateArray)
+    event = Event(device(Q.data))
+    event = begin_ghost_exchange!(Q; dependencies = event)
+    wait(device(Q.data), event)
+end
 
-        Q.recvreq[n] = MPI.Irecv!(
-            (@view transfer[:, Q.nabrtovmaprecv[n]]),
-            Q.nabrtorank[n],
-            Q.commtag,
-            Q.mpicomm,
-        )
-    end
+function __no_overlap_end_ghost_exchange!(Q::MPIStateArray)
+    event = Event(device(Q.data))
+    event = end_ghost_exchange!(Q; dependencies = event)
+    wait(device(Q.data), event)
 end
 
 """
-    start_ghost_exchange!(Q::MPIStateArray; dorecvs=true)
+    begin_ghost_exchange!(Q::MPIStateArray; dependencies = nothing)
 
-Start the MPI exchange of the data stored in `Q`. If `dorecvs` is `true` then
-`post_Irecvs!(Q)` is called, otherwise the caller is responsible for this.
-
-This function will fill the send buffer (on the device), copies the data from
-the device to the host, and then issues the send. Previous sends are waited on
-to ensure that they are complete.
+Begin the MPI halo exchange of the data stored in `Q`.  A KernelAbstractions
+`Event` is returned that can be used as a dependency to end the exchange.
 """
-function start_ghost_exchange!(Q::MPIStateArray; dorecvs = true)
+function begin_ghost_exchange!(Q::MPIStateArray; dependencies = nothing)
+    if !all(r -> r == MPI.REQUEST_NULL, Q.recvreq)
+        error("The currrent ghost exchange must end before another begins.")
+    end
 
-    dorecvs && post_Irecvs!(Q)
+    __Irecv!(Q)
 
     # wait on (prior) MPI sends
-    finish_ghost_send!(Q)
+    @tic mpi_sendwait
+    MPI.Waitall!(Q.sendreq)
+    fill!(Q.sendreq, MPI.REQUEST_NULL)
+    @toc mpi_sendwait
 
     @tic mpi_sendcopy
```

Comment:
This `tic` no longer measures anything but the time it takes to launch two kernels.

Reply:
Doesn't measure the time in `MPI.Waitall!()`?  I think all of the calls to `@tic` and `@toc` have this issue.  I suggest that this be addressed in a future pr that makes `@tic` and `@toc` KernelAbstractions and MPI aware.